Path: Added post processor for Fangling plasma cutter

nik570 opened the pull request:
Added post processor for Fangling plasma cutter to path workbench.
Fangling plasma cutters seems to be a rather common in Chinese plasma cutters and is probably used in many different brands. It have been tried using the engrave operation in demo mode on real hardware and seems to work but not yet cut real metal.


Thank you for creating a pull request to contribute to FreeCAD! Place an "X" in between the brackets below to "check off" to confirm that you have satisfied the requirement, or ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10) if there is something you don't understand.

- [X]  Your Pull Request meets the requirements outlined in section 5 of [CONTRIBUTING.md](https://github.com/FreeCAD/FreeCAD/blob/master/CONTRIBUTING.md) for a Valid PR

Please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [1.0 Changelog Forum Thread](https://forum.freecad.org/viewtopic.php?f=10&t=69438).

---

luzpaz commented:
Please amend commit and prepend `Path: ` to commit title. TIA

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9167"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_9167/pipeline.svg"/></a> for feature branch [PR_9167](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9167). Pipeline [826859766](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/826859766) was triggered at [9de771b](https://github.com/FreeCAD/FreeCAD/pull/9167/commits/9de771b019e4e1bbfe5aa2a8c646661461e1d495). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

benj5378 commented on src/Mod/Path/Path/Post/scripts/fangling_post.py:
```diff
@@ -0,0 +1,450 @@
+            # Now add the remaining parameters in order
+            for param in params:
+                if param in c.Parameters:
+                    if param == 'F' and (currLocation[param] != c.Parameters[param] or OUTPUT_DOUBLES):
+                        if c.Name not in ["G0", "G00"]:  # mach3_4 doesn't use rapid speeds
```
```suggestion
                if (
                    param == 'F' and
                    (currLocation[param] != c.Parameters[param] or OUTPUT_DOUBLES)
                    and c.Name not in ["G0", "G00"]  # mach3_4 doesn't use rapid speeds
                ):
```

benj5378 commented on src/Mod/Path/Path/Post/scripts/fangling_post.py:
```diff
@@ -0,0 +1,450 @@
+                                outstring.append(param + format(float(speed.getValueAs(UNIT_SPEED_FORMAT)), precision_string))
+                            else:
+                                continue
+                    elif param == 'T':
+                        outstring.append(param + str(int(c.Parameters['T'])))
+                    elif param == 'H':
+                        outstring.append(param + str(int(c.Parameters['H'])))
+                    elif param == 'D':
+                        outstring.append(param + str(int(c.Parameters['D'])))
+                    elif param == 'S':
+                        outstring.append(param + str(int(c.Parameters['S'])))
```
```suggestion
                elif param in ['T', 'H', 'D', 'S']:
                    outstring.append(param + str(int(c.Parameters[param])))
```
Also why `str(int(var))` and not just `str(var)`?

benj5378 commented on src/Mod/Path/Path/Post/scripts/fangling_post.py:
```diff
@@ -0,0 +1,450 @@
+                        outstring.append(param + str(int(c.Parameters['D'])))
+                    elif param == 'S':
+                        outstring.append(param + str(int(c.Parameters['S'])))
+                    else:
+                        if (not OUTPUT_DOUBLES) and (param in currLocation) and (currLocation[param] == c.Parameters[param]):
+                            continue
+                        else:
+                            pos = Units.Quantity(c.Parameters[param], FreeCAD.Units.Length)
+                            outstring.append(
+                                param + format(float(pos.getValueAs(UNIT_FORMAT)), precision_string))
```
```suggestion
                elif not OUTPUT_DOUBLES and param in currLocation and currLocation[param] == c.Parameters[param]:
                        continue
                else:
                    pos = Units.Quantity(c.Parameters[param], FreeCAD.Units.Length)
                    outstring.append(
                        param + format(float(pos.getValueAs(UNIT_FORMAT)), precision_string))
```

benj5378 commented on src/Mod/Path/Path/Post/scripts/fangling_post.py:
```diff
@@ -0,0 +1,450 @@
+                            outstring.append(
+                                param + format(float(pos.getValueAs(UNIT_FORMAT)), precision_string))
+
+            if adaptiveOp and c.Name in ["G0", "G00"]:
+                if opHorizRapid and opVertRapid:
```
```suggestion
            if adaptiveOp and c.Name in ["G0", "G00"] and opHorizRapid and opVertRapid:
```

benj5378 commented on src/Mod/Path/Path/Post/scripts/fangling_post.py:
```diff
@@ -0,0 +1,450 @@
+                print('M6 tool change removed')
+
+            if command == "message":
+                if OUTPUT_COMMENTS is False:
```
```suggestion
                if not OUTPUT_COMMENTS:
```

benj5378 reviewed (commented):
I hope you find my feedback helpful!

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9167"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_9167/pipeline.svg"/></a> for feature branch [PR_9167](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9167). Pipeline [829263138](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/829263138) was triggered at [82233f6](https://github.com/FreeCAD/FreeCAD/pull/9167/commits/82233f68cbb63b442a5191ca59edd977728c6f28). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9167"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_9167/pipeline.svg"/></a> for feature branch [PR_9167](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9167). Pipeline [830520578](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/830520578) was triggered at [6f80066](https://github.com/FreeCAD/FreeCAD/pull/9167/commits/6f80066c4be781b128a382541ec20787dcaa4511). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

luzpaz commented:
@nik570 ping

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9167"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_9167/pipeline.svg"/></a> for feature branch [PR_9167](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9167). Pipeline [831684308](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/831684308) was triggered at [65465ab](https://github.com/FreeCAD/FreeCAD/pull/9167/commits/65465ab35367d5a2975b96e9070a79fb40f9882c). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

benj5378 commented:
@nik570 Did you look at my feedback?

nik570 commented:
Yes I think the suggestions are good.

First time I do a pull request on Github so a little unsure how to get things done.

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9167"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_9167/pipeline.svg"/></a> for feature branch [PR_9167](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9167). Pipeline [832251002](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/832251002) was triggered at [a7dabae](https://github.com/FreeCAD/FreeCAD/pull/9167/commits/a7dabae63af7ccef62273fbc3996912cc041fac0). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9167"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_9167/pipeline.svg"/></a> for feature branch [PR_9167](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9167). Pipeline [832252078](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/832252078) was triggered at [d56291f](https://github.com/FreeCAD/FreeCAD/pull/9167/commits/d56291f26a3200b4085e128031e6b878b8704224). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9167"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_9167/pipeline.svg"/></a> for feature branch [PR_9167](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9167). Pipeline [832253012](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/832253012) was triggered at [93e6eec](https://github.com/FreeCAD/FreeCAD/pull/9167/commits/93e6eeca6dc133379a3c60b58bc0b5cdded52107). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

nik570 commented:
After the changes you suggested script does not work anymore. Nor sure why, found and solved one problem with the indentation.

unning the Python command 'Path_Post' failed:
Traceback (most recent call last):
  File "/home/karlnick/Prog/freecad_fangling-build/Mod/Path/Path/Post/Command.py", line 588, in Activated
    result, gcode, name = self.exportObjectsWith(sublist, partname, job, idx)
  File "/home/karlnick/Prog/freecad_fangling-build/Mod/Path/Path/Post/Command.py", line 523, in exportObjectsWith
    gcode = processor.export(objs, filename, postArgs)
  File "/home/karlnick/Prog/freecad_fangling-build/Mod/Path/Path/Post/Processor.py", line 80, in export
    return self.script.export(obj, filename, args)
  File "/home/karlnick/Prog/freecad_fangling-build/Mod/Path/Path/Post/scripts/fangling_post.py", line 237, in export
    gcode += parse(obj)

can only concatenate str (not "NoneType") to str

benj5378 commented:
@nik570 

Yeah, I would advice you to not use the "Use suggestions" button but reincorporate them manually. Also, don't rely on my suggestions being correct - maybe there are reasons behind how it is written that I didn't see. We need to fix the 15 commits in this PR so there are only 2: the easiest way to fix them is if I can guide you live, could you please join https://discord.gg/uh85ZRNcfk (if you are available now)

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9167"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_9167/pipeline.svg"/></a> for feature branch [PR_9167](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9167). Pipeline [832508968](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/832508968) was triggered at [9a18563](https://github.com/FreeCAD/FreeCAD/pull/9167/commits/9a185635feb769d13d8309ce1893b2f572e15504). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

nik570 commented:
Found the problem and seems to work again.

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9167"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_9167/pipeline.svg"/></a> for feature branch [PR_9167](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9167). Pipeline [832534024](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/832534024) was triggered at [42abde0](https://github.com/FreeCAD/FreeCAD/pull/9167/commits/42abde0396c2a13e026332f5327446125a4a5878). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

benj5378 commented:
> We need to fix the 15 commits in this PR so there are only 2: the easiest way to fix them is if I can guide you live, could you please join https://discord.gg/uh85ZRNcfk (if you are available now)

@nik570 Please don't merge master into your branch all the time. This is what we actually want to remove again

nik570 commented:
Only here intermittently. If needed I could join discord another day otherwise if you have instructions I do it myself.

Should not press the update branch button anymore.

benj5378 commented:
@nik570 Just Discord for chatting back and forth rather than an actual call. I was thinking about guiding you how to first delete all the merge into master commits and then rebase on top of the master. Though in your case, it might be easier to make a new branch off master and then cherry picking the your main commit [Added post processor for Fangling plasma cutter to path workbench.](https://github.com/FreeCAD/FreeCAD/pull/9167/commits/12837580c326fa72c43f22c5145dd761927dee30) and committing immidately. Then afterwards cherry picking all your [Update src/Mod/Path/Path/Post/scripts/fangling_post.py](https://github.com/FreeCAD/FreeCAD/pull/9167/commits/93e6eeca6dc133379a3c60b58bc0b5cdded52107) commits and commit amend after you cherry picked them all.

Do you use git in terminal or a GUI like SourceTree? (can highly recommend GitKraken)

nik570 commented:
Use git on command line or integrated into development environment. Will do as I suggest but are probably busy with other things both today and tomorrow.

nik570 commented:
Have done as you said, deleted old commits and added new.

It is a first implementation. Adding kerf compensation would be good and maybe some code cleanup. Dressup to add start and maybe exit move would also be good but this should not be part of post processor.

G-code is simple, reading the manual:
  G99 Parameter: X/U Y/V I J Part option parameter
  G92 Parameter: X Y Reference point setting
  G91 / G90 No parameter Relative/Absolute coordinate
  G20 / G21 No parameter Inches/metric
  G41 / G42 No parameter Left/right Kerf compensation
  G40 No parameter Cancel Kerf compensation
  G00 Parameter: X/U Y/V straight line rapid moving tool
  G01 Parameter: X/U Y/V Straight line cutting
  G02 Parameter: X/U Y/V I J Clockwise arc cutting
  G03 Parameter: X/U Y/V I J Anticlockwise arc cutting
  G04 Parameter: P Time delay

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9167"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_9167/pipeline.svg"/></a> for feature branch [PR_9167](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9167). Pipeline [837696986](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/837696986) was triggered at [224fee8](https://github.com/FreeCAD/FreeCAD/pull/9167/commits/224fee8e64f63f54c84f1645f393a2cca2ed880a). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

sliptonic commented:
> @luzpaz Can you find an reviewer for this? Preferably someone currently active.

I'm watching this PR.  It looks like you gave a very nice and comprehensive code review. Thank you.  Some of the conversations are still unresolved.  @nik570,  please resolve the outstanding points, either by modifying code or by commenting and choosing not to.

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9167"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_9167/pipeline.svg"/></a> for feature branch [PR_9167](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9167). Pipeline [838762967](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/838762967) was triggered at [fee5c55](https://github.com/FreeCAD/FreeCAD/pull/9167/commits/fee5c551ff5d6ff00accdae46ccf5d272fde7a42). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

benj5378 commented:
@nik570 Congratulations with your first PR to FreeCAD! Please come back for more - if you don't know what to work on, I am sure we can find something.

luzpaz commented:
Indeed, congrats @nik570! And thanks to your input @benj5378 :+1:
